psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java

```diff
@@ -2,130 +2,135 @@
+                converter = EnclaveConstraintConverter.class)
+        private EnclaveConstraint constraint;
+
+    /**
+     * Use Kryo to serialize inputs from client to enclave
+     */
+    private Output serializeMessage(InputData listOfCreditCardNumbers){
+        Kryo kryo = new Kryo();
+        Output output = new Output(new ByteArrayOutputStream());
+        kryo.register(InputData.class, new InputDataSerializer());
+        kryo.writeObject(output, listOfCreditCardNumbers);
+        output.close();
+        return output;
+    }
+
+    @Override
+    public Void call() throws Exception {
+        //a new private key is generated. Enclave Client is created using this private key and constraint.
+        //a corresponding public key will be used by the enclave to encrypt data to be sent to this client
+        EnclaveClient enclaveClient = new EnclaveClient(constraint);
+
+        try (WebEnclaveTransport transport = new WebEnclaveTransport(url);
+             EnclaveClient client = enclaveClient)
+        {
+            //retrieve the enclaveInstanceInfo object, i.e. the remote attestation object and verify it against the
+            //constraint
+            client.start(transport);
+
+            //collect merchants and service providers credit card numbers list
+            InputData inputData = getInputData();
+
+            byte[] requestMailBody = serializeMessage(inputData).getBuffer();
+
+            //client will send its credit card list to enclave and wait for other client to send their list
+            EnclaveMail responseMail = client.sendMail(requestMailBody);
+
+            //responseMail is null till enclave doesn't reply back to the client
+            if(responseMail == null) {
+                do
+                {
+                    //poll for reply to enclave
+                    responseMail = enclaveClient.pollMail();
+                    Thread.sleep(10000);
+                } while (responseMail==null);
             }
+            System.out.println("Ad Conversion Rate is : " + new String(responseMail.getBodyAsBytes()));
         }
+        return null;
+    }
 
-        //take inputs from user
+    private InputData getInputData() {
         InputData inputData = new InputData();
         List<UserDetails> userDetailsList = null;
-        List<AdDetails> adDetailsList= null;
+        List<AdDetails> adDetailsList = null;
 
-        inputData.setClientType(clientType);
-        if(MERCHANT.equals(clientType)) {
-            userDetailsList = new ArrayList<>(args.length);
-            for (int i =2; i< args.length; i++) {
-                UserDetails userDetails = new UserDetails(args[i]);
+        inputData.setClientType(creditCardNumbers.get(0));
+
+        System.out.println("string.length" + creditCardNumbers.size());
+        System.out.println("string.length" + creditCardNumbers.get(0));
+
+        if (MERCHANT.equals(creditCardNumbers.get(0))) {
```

Comment:
It will be better and clearer to have a separate flag called `--role` (or whatever term is best) for the merchant/service provider difference. You can define an enum for it an picocli will automatically deal with it.
```
enum class Role {
  MERCHANT,
  SERVICE_PROVIDER
}
```

Having it in the credit card numbers list will be confusing to the developer

Reply:
done